File: src/main/java/org/blacksun/graph/channel/HalfDuplexChannel.java
package org.blacksun.graph.channel;

import org.blacksun.graph.node.GraphNode;

public class HalfDuplexChannel extends AbstractChannel {
    private HalfDuplexChannel pair;

    private HalfDuplexChannel(HalfDuplexChannel channel) {
        super(channel.toNode, channel.fromNode, channel.weight, channel.errors);
        pair = channel;
    }

    public HalfDuplexChannel(GraphNode fromNode, GraphNode toNode, int weight, double errors) {
        super(fromNode, toNode, weight, errors);
        pair = new HalfDuplexChannel(this);
        connect();
    }

    @Override
    public void setUsed(boolean used) {
        this.used = used;
        pair.used = used;
    }

    @Override
    public void connect() {
        super.connect();
        toNode.addConnection(pair);
    }

    @Override
    public void remove() {
        super.remove();
        toNode.removeConnection(pair);
    }

    @Override
    protected String getDirection() {
        return "<-->";
    }
}

File: src/main/java/org/blacksun/graph/channel/HalfDuplexChannel.kt
package org.blacksun.graph.channel

import org.blacksun.graph.node.GraphNode

class HalfDuplexChannel : AbstractChannel {
    private var pair: HalfDuplexChannel

    override var isUsed = false
        set(used) {
            field = used
            if (pair.isUsed != used)
                pair.isUsed = used
        }

    override val direction = "<-->"

    private constructor(channel: HalfDuplexChannel) : super(channel.toNode, channel.fromNode, channel.weight, channel.errors) {
        pair = channel
    }

    constructor(fromNode: GraphNode, toNode: GraphNode, weight: Int, errors: Double) : super(fromNode, toNode, weight, errors) {
        pair = HalfDuplexChannel(this)
        connect()
    }

    override fun connect() {
        super.connect()
        toNode.addConnection(pair)
    }

    override fun remove() {
        super.remove()
        toNode.removeConnection(pair)
    }
}
